waku/nwaku.go

```diff
@@ -532,12 +533,70 @@ func New(nwakuCfg *WakuConfig, logger *zap.Logger) (*Waku, error) {
 	}, nil
 }
 
+// The event callback sends back the node's ctx to know to which
+// node is the event being emited for. Since we only have a global
+// callback in the go side, We register all the nodes that we create
+// so we can later obtain which instance of `WakuNode` is should
+// be invoked depending on the ctx received
+
+var nodeRegistry map[unsafe.Pointer]*WakuNode
+
+func init() {
+	nodeRegistry = make(map[unsafe.Pointer]*WakuNode)
+}
+
+func registerNode(node *WakuNode) {
+	_, ok := nodeRegistry[node.wakuCtx]
+	if !ok {
+		nodeRegistry[node.wakuCtx] = node
+	}
+}
+
+func unregisterNode(node *WakuNode) {
+	delete(nodeRegistry, node.wakuCtx)
+}
+
 //export globalEventCallback
 func globalEventCallback(callerRet C.int, msg *C.char, len C.size_t, userData unsafe.Pointer) {
-	// This is shared among all Golang instances
-	// TODO-nwaku
-	// self := Waku{wakuCtx: userData}
-	// self.MyEventCallback(callerRet, msg, len)
+	if callerRet == C.RET_OK {
+		eventStr := C.GoStringN(msg, C.int(len))
+		node, ok := nodeRegistry[userData]
```

Comment:
May add a comment explaining what info `userData` contains in this point?

Reply:
Ok I got confused and was hard to follow, but I think I got it now

This comment says that the ctx is returned by `WakuCallback` (in the `userData` field) to know to which node an event corresponds
https://github.com/waku-org/waku-go-bindings/blob/8e5ed15dbd0397885fa5ab6340c0cd3ce9ef5b1b/waku/nwaku.go#L536-L542

We can see how we send the context twice over here (one in the `ctx` field and the other in the `userData` one)

https://github.com/waku-org/waku-go-bindings/blob/8e5ed15dbd0397885fa5ab6340c0cd3ce9ef5b1b/waku/nwaku.go#L98-L112

So when calling the event callback, libwaku will have the original context in `eventUserData` and will send it back [here](https://github.com/waku-org/nwaku/blob/ab0c1d4aa0e545c2c74b7e1ea8045823370eee4d/library/libwaku.nim#L76-L79), and we'll know to which node the event corresponds

Uff this is actually pretty hacky and hard to follow, but maybe I'm missing something. I get particularly confused by our use of `userData`, which I understand it was also some other hacky solution to working with Rust's closures.

And I see we also use the `userData` field to hand over a pointer with memory allocated for a response, and now in this case to send back the context.

Idk, I get the idea that the `userData` field is used for many unrelated things and it got pretty messy. Everything seems to work but at some point we should revisit the design and use of this field.

@richard-ramos @Ivansete-status did I understand it well?